### PR TITLE
Realign spec and proofs

### PR DESCRIPTION
**Pull request by blaxill**
A few admits still left to go.

### CODE REVIEW

File: cava2/FifoProperties.v
Comment (blaxill) [inline cava2/FifoProperties.v:903]: @jadephilipoom can I add these globally or do you think its a bad idea?

Comment (jadephilipoom) [reply cava2/FifoProperties.v:903]: I think it's fine to add them, but in the `If.v` file rather than this one.

**Comment by samuelgruetter** [comment]
I see a lot of proofs here that start with `N.bits_inj` and then prove that for each bit index, `N.testbit` at that index returns the same on the left and right of the equality to prove.
I have already developed automation for this proof strategy twice (but for `Z`, not for `N`):
* A [first approach](https://github.com/mit-plv/coqutil/blob/c4265ae220b6ce8dcb0d0f1a90fbc38169ce788b/src/coqutil/Z/prove_Zeq_bitwise.v), mostly used in riscv-coq, was getting a bit complex, and ran into performance problems in `lia`, `btauto`, and rewriting (though `rewrite_strat` helped), and still fails to prove some goals I think it should prove, and also runs forever if I add all the rewrite lemmas I think it should use.
* A simpler [second approach](https://github.com/mit-plv/coqutil/blob/c4265ae220b6ce8dcb0d0f1a90fbc38169ce788b/src/coqutil/Z/bitblast.v), which is easier to deal with, but less powerful than the first one.

I don't claim that there's anything ready to use for you there, but depending on how many more bitwise goals you need to prove, maybe you could harvest some reusable snippets from these files.
Or if you expect really many bitwise goals, it might be worth considering developing a tactic that's actually good and works for all goals in cava & riscv-coq & bedrock2, but that would be a bit of an investment...

**Comment by blaxill** [comment]
> I see a lot of proofs here that start with `N.bits_inj` and then prove that for each bit index, `N.testbit` at that index returns the same on the left and right of the equality to prove.
> I have already developed automation for this proof strategy twice (but for `Z`, not for `N`):
> 
> * A [first approach](https://github.com/mit-plv/coqutil/blob/c4265ae220b6ce8dcb0d0f1a90fbc38169ce788b/src/coqutil/Z/prove_Zeq_bitwise.v), mostly used in riscv-coq, was getting a bit complex, and ran into performance problems in `lia`, `btauto`, and rewriting (though `rewrite_strat` helped), and still fails to prove some goals I think it should prove, and also runs forever if I add all the rewrite lemmas I think it should use.
> * A simpler [second approach](https://github.com/mit-plv/coqutil/blob/c4265ae220b6ce8dcb0d0f1a90fbc38169ce788b/src/coqutil/Z/bitblast.v), which is easier to deal with, but less powerful than the first one.
> 
> I don't claim that there's anything ready to use for you there, but depending on how many more bitwise goals you need to prove, maybe you could harvest some reusable snippets from these files.
> Or if you expect really many bitwise goals, it might be worth considering developing a tactic that's actually good and works for all goals in cava & riscv-coq & bedrock2, but that would be a bit of an investment...

Thanks. I had previously approached those proofs with normal rewriting, but it was taking a huge number of rewrites to massage the goal and was taking too long. The bitwise specialised-for-my-specific goal tactics didn't actually take too long, but I'll definitely use your machinery in the future

**Comment by jadephilipoom** [comment]
I've already got a tactic for bitwise reasoning in `N` in our repo called `push_Ntestbit`. Just apply `N.bits_inj`, then `intro`, then `push_Ntestbit` will take you as far as it can go. Might need further `destruct`s to rewrite with lemmas that have arithmetic side conditions (e.g. for `N.testbit (N,ones x) i` you will need to destruct `i <? x`), so there's room for more automation there, but should help with the majority of the manual steps.

**Comment by blaxill** [comment]
The proof is reduced to some admitted lemmas about `concat_bytes` which all look to be true- I'd like to address them in a follow up PR.

I should have broken down the main proofs further/better but oh well I've learnt my lesson for the future.

@jadephilipoom could you take a look please?

File: cava2/FifoProperties.v
Comment (jadephilipoom) [inline cava2/FifoProperties.v:865]: Looking at the invariant and spec, it seems like this value isn't really used; the spec determines whether the output is valid just based on the state data, and it's not mentioned in the invariant except in a clause that says it matches `out_valid`, which is also not referenced elsewhere. I think you could probably just remove it, which would simplify the reasoning a bit in higher-level proofs.

File: cava2/FifoProperties.v
Comment (jadephilipoom) [inline cava2/FifoProperties.v:713]:
```suggestion
  Lemma length_N_to_bytes n xs:
```
I actually have this proven in a local branch that I'll merge once its dependencies are merged, so no need to worry about it.

File: cava2/FifoProperties.v
Comment (jadephilipoom) [inline cava2/FifoProperties.v:704]: I think these could be reduced to a single lemma about `N.testbit (BigEndianBytes.concat_bytes _)`, something like:
```coq
N.testbit (BigEndianBytes.concat_bytes bs) n = N.testbit (Byte.to_N (nth (n / 8) bs 0)) (n mod 8)
```
You could then use lemmas about `nth` to do the rest.

File: cava2/FifoProperties.v
Comment (jadephilipoom) [inline cava2/FifoProperties.v:673]: I think this one should be true in general:
```suggestion
  Lemma bytes_to_N_id xs:
    BigEndianBytes.N_to_bytes n (BigEndianBytes.concat_bytes (List.resize Byte.x00 n xs)) = xs.
  Admitted.
```

**Review by jadephilipoom — APPROVED** [review]
A few comments for small improvements/cleanup, but I think it's OK to merge without blocking on them.

File: cava2/FifoProperties.v
Comment (blaxill) [reply cava2/FifoProperties.v:434]: The testbit of the constant doesn't reduce with push_Ntestbit alone, but I've reduced it to 

```
450     destr (n' <? 32)%nat.
451     { do 32 (destruct n'; [push_Ntestbit; now boolsimpl| ]); lia. } 
452     {  
453       push_Ntestbit; boolsimpl. 
454       rewrite N.sub_add by lia. 
455       repeat rewrite (N.testbit_high_lt _ _ _ H) by lia.  
456       now boolsimpl.   
457     }
```

It doesn't matter of course, but its a noticeably slower, presumably because of the rewriting order of push_Ntestbit.

File: cava2/FifoProperties.v
Comment (blaxill) [reply cava2/FifoProperties.v:865]: It's used in the first branch of `if negb consumer_ready then (latched_valid,..`. Unfortunately in that case it can't be recovered from `state_data`.

**Pull request merged**
